docs/proposals/affinity.md

```diff
@@ -0,0 +1,341 @@
+}
+
+type AffinityTerm struct {
+	PodSelector *PodSelector `json:"podSelector,omitempty"` // empty PodSelector matches all pods
+	TopologyKey string `json:"topologyKey,omitempty"`
+}
+```
+
+One caveat: As noted above, the normal semantics of empty `PodSelector` is to match all pods.
+However, `PodSelector` is scoped to namespace, therefore the normal semantics do not allow
+us to match all pods globally. We need to special-case interpret empty `PodSelector` in this
+context to mean "all pods globally."
+
+To explain how this API works, let's say that the `PodSpec` of a pod `P` has an `AffinitySelector`
+that is configured as follows:
+
+```go
+AffinitySelector {
+	HardAffinityExpressions: {{PodSelector: P1, TopologyKey: "node"}},
+	HardAntiAffinityExpressions: {{PodSelector: P2, TopologyKey: "rack"}},
+	SoftAffinityExpressions: {{PodSelector: P3, TopologyKey: "zone"}},
+	SoftAntiAffinityExpressions: {{PodSelector: P4, TopologyKey: "power"}}
+}
+```
+
+Then when scheduling pod P, the scheduler
+* Can only schedule P onto nodes that are running pods that satisfy `P1`. (Assumes all nodes have a label with key "node" and value specifying their node name.)
+* Cannot schedule P onto any racks that are running pods that satisfy `P2`. (Assumes all nodes have a label with key "rack" and value specifying their rack name.)
+* Should try to schedule P onto zones that are running pods that satisfy `P3`. (Assumes all nodes have a label with key "zone" and value specifying their zone.)
+* Should try not to schedule P onto any power domains that are running pods that satisfy `P4`. (Assumes all nodes have a label with key "power" and value specifying their power domain.)
+
+When `HardAffinityExpressions` has multiple elements, the requirements are ANDed. Likewise for `HardAntiAffinityExpressions`.
+For `SoftAffinityExpressions` and `SoftAntiAffinityExpressions`, nodes that satisfy more of the rules are preferred over those that satisfy fewer.
+
+## A comment on symmetry
+
+One thing that makes affinity and anti-affinity tricky is symmetry.
+
+Imagine a cluster that is running pods from two services, S1 and S2. Imgine that the pods of S1 have a hard anti-affinity rule
+"do not run me on nodes that are running pods from S2." It is not sufficient just to check that there are no S2 pods on a node when
+you are scheduling a S1 pod. You also need to ensure that there are no S1 pods on a node when you are scheduling a S2 pod,
+*even though the S2 pod does not have any anti-affinity rules*. Otherwise if an S1 pod schedules before an S2 pod, the S1
+pod's hard anti-affinity rule can be violated by a later-arriving S2 pod.
+
+Note that while hard anti-affinity is symmetric,
+hard affinity is *not* symmetric. That is, if the pods of S1 have a hard affinity rule "run me on nodes that are running
+pods from S2," it is not required that there be S1 pods on a node in order to schedule a S2 pod onto that node.
+
```

Comment:
One might argue that Pods themselves give us proper node-level hard affinity?  I don't actually think that's true (i.e. forcing a pod that wants to be on the same node as another pod to merge with that pod definition) doesn't work out well in many cases), but I think it's worth mentioning here.  Clearly Pods only provide the notion of hard node affinity, and don't generalize to non-node.